### DISCUSSION TITLE
What is the purpose for subnet intersection checking in netutil.go?

### DISCUSSION THREAD

**Discussion by compdude22** [containerd/nerdctl#3430] [Q&A]
(occasional lurker, first time poster)

I'm trying to understand what the purpose of the [subnet intersection/overlap check in netutil.go](https://github.com/yldoge/nerdctl/blob/d92bdf6342fe937400efa3d57eb74444599cc3da/pkg/netutil/netutil.go#L493), as part of the parseSubnet() function.

As background info, I am using `nerdctl compose` to run my homelab's network services, but I intend for all containers bridged directly to my LAN and sharing the existing subnet, rather than using NPTv6 or NAT. Since my LAN has a standard /64 subnet, I've statically carved out a small range within that subnet so that a CNI IPAM plugin can locally assign those addresses to containers on-the-fly, without having to perform SLAAC or DHCPv6.

Bridging to the LAN is working fine using the `macvlan` CNI driver. But I am unable to get nerdctl to create the necessary IPAM config, even though I can create the equivalent CNI config manually. The issue appears to arise when configuring an IPAM with a subnet that overlaps with a subnet already known to the host's network namespace. In my case, I'm trying to create an IPAM config that exactly mirrors the physical LAN, so I am getting error messages of the form `<subnet> overlaps with other one on this address space` and the network is not created.

Example setup:
```
Host address: 2001:db8:42:101
LAN subnet: 2001:db8:42::/64
Range of addresses dedicated to containers: 2001:db8:42:cccc::/112

$ nerdctl network create --driver=macvlan --subnet=2001:db8:42::/64 --ip-range=2001:db8:42:cccc::/112 mybridge
FATA[0000] subnet 2001:db8:42::/64 overlaps with other one on this address space
```

The `ip-range=` isn't actually needed to produce this error, since it appears to just be the `subnet` value that's problematic, as the subnet already appears in the host network namespace's routing table. At least, that's what my cursory read of the code suggests.

What I'm trying to understand is what this subnet intersection check is seeking to protect against. Unless I'm mistaken, it should not matter is two container networks are assigning addresses from the same subnet, or even if two containers are assigned the exact same address, so long as they are on separate networks which do not bridge (at L2) or route (at L3) to each other.

In my case, I am intentionally bridging my containers to the same network as the host, precisely to have a "flat" network. Which is to say, the intersection of my container network and my physical network is a circle. But this check prevents me from doing so via `nerdctl network create`.

The check would also prevent someone with two distinct physical L2 networks from being able to use the same L3 addressing for containers assigned to either L2 network. There is no reason to deny the same L3 subnet if it will be used on entirely-separated L2 networks.

My use-case only covers IPv6 because I operate an IPv6-only network. But I think the same would arise with IPv4 networks. So I'm now looking to understand the rationale for this check, which I can vaguely understand as a warning for network novices. But ultimately, the underlying CNI IPAM plugin doesn't seem to have any issue with assigning addresses the way I desire above, so I wish to understand why `nerdctl network` does. Or perhaps the check and error message can be removed outright.

The two prior GitHub discussions with keywords "subnet overlaps" don't really provide additional color, with the error message not being a causal factor to their questions.

**Comment by AkihiroSuda** [comment]
This was added in:
- https://github.com/containerd/nerdctl/pull/1245

**Comment by compdude22** [comment]
@AkihiroSuda , I do see that the change was introduced in that change. But I'm not understanding why the logic is checking against *all* subnets known to the system -- from GetLiveNetworkSubnets() -- rather than just the subnets used/created by `nerdctl network create`.

It seems like the behavior would work just fine without having to involve the system's non-nerdctl subnets, if that's something which could be tracked or listed.

**Reply by AkihiroSuda** [reply]
cc @junnplus Do you remember the reason?

**Reply by junnplus** [reply]
Sorry, I don’t quite remember the exact reasoning behind that change.

**Reply by compdude22** [reply]
Would there be any objection to removing the  check outright?

**Reply by AkihiroSuda** [reply]
No objection from me

**Reply by eric-ela-b3006c** [reply]
I would also be quite happy with the removal of the check

**Reply by compdude22** [reply]
What would be the procedure to move forward with removal of the check? Should an Issue be opened first? Were it not for preexisting commitments, I would open a PR to make this three-line change.

**Reply by AkihiroSuda** [reply]
Feel free to just submit a PR directly

**Comment by eric-ela-b3006c** [comment]
A mention to this issue was added in #4082

**Comment by Yash121l** [comment]
The check is `parseSubnet` in `pkg/netutil/netutil.go`, which runs `IntersectsWithNetworks` against `usedSubnets()`: every address currently on the host's interfaces plus the subnets of CNI configs already in the netconfpath. It came in with #1245, whose real goal was automatic allocation: `GetFreeSubnet` walks up from 10.4.1.0/24 until nothing intersects, and the same predicate was reused to validate an explicit `--subnet`. So it is an allocator collision guard written for the bridge driver and applied to every driver, with no awareness that macvlan and ipvlan are meant to sit on the host's own segment. You can see the bridge shape in `NetworkConfig.subnets` in `netutil_unix.go`, which only reports subnets for `bridge` plugins, so nerdctl's own macvlan networks never feed the used set. There is no flag to relax it; the workarounds are creating the network while the parent interface carries no address, or writing the conflist yourself into `--cni-netconfpath`, which nerdctl lists and runs against.
